File: app/src/main/java/com/arcinteractive/spaces/data/model/Space.kt
package com.arcinteractive.spaces.data.model

import androidx.compose.ui.graphics.Color

data class Space(
    val id: String,
    val name: String,
    val emoji: String,
    val colorHex: String,
    val description: String,
    val template: SpaceTemplate,
    val ownerId: String,
    val memberIds: List<String>,
    val unreadCount: Int?,
    val enabledModules: List<SpaceModule>
) {
    val subtitle: String
        get() = description

    val modules: List<SpaceModule>
        get() = (enabledModules + SpaceModules.Settings).distinctBy { it.id }

    val filesEnabled: Boolean
        get() = enabledModules.any { it.id == SpaceModules.Files.id }

    val pollsEnabled: Boolean
        get() = enabledModules.any { it.id == SpaceModules.Polls.id }
}

enum class SpaceTemplate(val title: String, val subtitle: String, val suggestedEmoji: String) {
    Family("Family", "Shared updates and planning", "\uD83C\uDFE1"),
    Friends("Friends", "Chats, plans, and photos", "\uD83E\uDEE6"),
    Business("Business", "Projects and team coordination", "\uD83D\uDCBC"),
    Community("Community", "Events and announcements", "\uD83C\uDF0E"),
    Custom("Custom", "Start with a blank space", "✨");

    val defaultStatus: String
        get() = when (this) {
            Family -> "Family updates and plans"
            Friends -> "Conversations and hangouts"
            Business -> "Team communication hub"
            Community -> "Events and announcements"
            Custom -> "A new shared space"
        }

    val defaultEnabledModules: List<SpaceModule>
        get() = when (this) {
            Family -> listOf(SpaceModules.General, SpaceModules.Photos, SpaceModules.Events, SpaceModules.Members)
            Friends -> listOf(SpaceModules.General, SpaceModules.Photos, SpaceModules.Events, SpaceModules.Members)
            Business -> listOf(SpaceModules.General, SpaceModules.Files, SpaceModules.Events, SpaceModules.Members)
            Community -> listOf(SpaceModules.General, SpaceModules.Files, SpaceModules.Events, SpaceModules.Members)
            Custom -> listOf(SpaceModules.General, SpaceModules.Members)
        }
}

File: iOS/Spaces V2/Models/Space.swift
import Foundation

struct Space: Identifiable, Hashable {
    let id: String
    let name: String
    let emoji: String
    let tintHex: String
    let description: String
    let template: SpaceTemplate
    let ownerId: String
    let memberIds: [String]
    let unreadCount: Int?
    let enabledModules: [SpaceModule]
    let moduleOrder: [SpaceModule]

    var subtitle: String {
        description
    }

    var modules: [SpaceModule] {
        let enabledSet = Set(enabledModules + [.settings])
        return moduleOrder.filter { enabledSet.contains($0) }
    }

    var eventsEnabled: Bool {
        enabledModules.contains(.events)
    }

    var filesEnabled: Bool {
        enabledModules.contains(.files)
    }

    var pollsEnabled: Bool {
        enabledModules.contains(.polls)
    }
}

enum SpaceModule: String, CaseIterable, Identifiable, Hashable {
    case general = "general"
    case photos = "photos"
    case files = "files"
    case polls = "polls"
    case events = "events"
    case members = "members"
    case settings = "settings"

    var id: String { rawValue }

    var title: String {
        switch self {
        case .general: "Space Pings"
        case .photos: "Photos"
        case .files: "Files"
        case .polls: "Polls"
        case .events: "Events"
        case .members: "Members"
        case .settings: "Settings"
        }
    }

    var icon: String {
        switch self {
        case .general: "bubble.left.and.bubble.right.fill"
        case .photos: "photo.on.rectangle.angled"
        case .files: "folder.fill"
        case .polls: "chart.bar.xaxis"
        case .events: "calendar"
        case .members: "person.2.fill"
        case .settings: "gearshape.fill"
        }
    }

    var emoji: String {
        switch self {
        case .general: "💬"
        case .photos: "📷"
        case .files: "📁"
        case .polls: "📊"
        case .events: "📅"
        case .members: "👥"
        case .settings: "⚙️"
        }
    }

    var description: String {
        switch self {
        case .general: "Private conversations and quick check-ins"
        case .photos: "Shared photos and memes"
        case .files: "Shared documents and uploads"
        case .polls: "Questions and voting"
        case .events: "Plans and calendar"
        case .members: "People in this Space"
        case .settings: "Space preferences"
        }
    }

    static var requiredModules: [SpaceModule] {
        [.general, .photos, .members]
    }

    static var optionalModules: [SpaceModule] {
        [.events, .files, .polls]
    }

    static var configurableModules: [SpaceModule] {
        requiredModules + optionalModules
    }

    static var allModules: [SpaceModule] {
        configurableModules + [.settings]
    }
}

enum SpaceTemplate: String, CaseIterable, Identifiable {
    case family = "Family"
    case friends = "Friends"
    case business = "Business"
    case community = "Community"
    case custom = "Custom"

    var id: String { rawValue }

    var subtitle: String {
        switch self {
        case .family: return "Shared updates and planning"
        case .friends: return "Chats, photos, and plans"
        case .business: return "Projects and coordination"
        case .community: return "Announcements and events"
        case .custom: return "Start from a blank space"
        }
    }

    var suggestedEmoji: String {
        switch self {
        case .family: return "🏡"
        case .friends: return "🫶"
        case .business: return "💼"
        case .community: return "🌎"
        case .custom: return "✨"
        }
    }

    var defaultSubtitle: String {
        switch self {
        case .family: return "Family updates and plans"
        case .friends: return "Conversations and hangouts"
        case .business: return "Team communication hub"
        case .community: return "Events and announcements"
        case .custom: return "A new shared space"
        }
    }

    var defaultEnabledModules: [SpaceModule] {
        switch self {
        case .family:
            return SpaceModule.requiredModules + [.events]
        case .friends:
            return SpaceModule.requiredModules + [.events]
        case .business:
            return SpaceModule.requiredModules + [.events, .files]
        case .community:
            return SpaceModule.requiredModules + [.events, .files]
        case .custom:
            return SpaceModule.requiredModules
        }
    }

    var defaultModuleOrder: [SpaceModule] {
        let enabled = defaultEnabledModules
        let disabledOptional = SpaceModule.optionalModules.filter { !enabled.contains($0) }
        return enabled + disabledOptional + [.settings]
    }
}
